processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java

```diff
@@ -417,15 +424,104 @@ public Object apply(final Result<TopNResultValue> input)
   }
 
   @Override
-  public QueryRunner<Result<TopNResultValue>> preMergeQueryDecoration(QueryRunner<Result<TopNResultValue>> runner)
+  public QueryRunner<Result<TopNResultValue>> preMergeQueryDecoration(final QueryRunner<Result<TopNResultValue>> runner)
   {
-    return intervalChunkingQueryRunnerDecorator.decorate(runner, this);
+    return intervalChunkingQueryRunnerDecorator.decorate(
+        new QueryRunner<Result<TopNResultValue>>()
+        {
+          @Override
+          public Sequence<Result<TopNResultValue>> run(
+              Query<Result<TopNResultValue>> query, Map<String, Object> responseContext
+          )
+          {
+            if (!(query instanceof TopNQuery)) {
+              return runner.run(query, responseContext);
+            } else {
+              final TopNQuery topNQuery = (TopNQuery) query;
+              if (TopNQueryEngine.canApplyExtractionInPost(topNQuery)) {
+                final DimensionSpec dimensionSpec = topNQuery.getDimensionSpec();
+                return runner.run(
+                    new TopNQuery(
+                        topNQuery.getDataSource(),
+                        new DefaultDimensionSpec(dimensionSpec.getDimension(), dimensionSpec.getOutputName()),
+                        topNQuery.getTopNMetricSpec(),
+                        topNQuery.getThreshold(),
+                        topNQuery.getQuerySegmentSpec(),
+                        topNQuery.getDimensionsFilter(),
+                        topNQuery.getGranularity(),
+                        topNQuery.getAggregatorSpecs(),
+                        topNQuery.getPostAggregatorSpecs(),
+                        topNQuery.getContext()
+                    )
+                    , responseContext
+                );
+              } else {
+                return runner.run(query, responseContext);
+              }
+            }
+          }
+        }
+        , this
+    );
   }
 
   @Override
   public QueryRunner<Result<TopNResultValue>> postMergeQueryDecoration(final QueryRunner<Result<TopNResultValue>> runner)
   {
-    return new ThresholdAdjustingQueryRunner(runner, config.getMinTopNThreshold());
+    final ThresholdAdjustingQueryRunner thresholdRunner = new ThresholdAdjustingQueryRunner(runner, config.getMinTopNThreshold());
```

Comment:
Please don't implement and expand on the logic as it exists in these methods.  I recently documented the QueryToolChest interfaces and realized that both the post and pre merge methods are entirely superfluous and just confusing.  There are recommendations in the docs to not implement these and instead do more work in the `mergeResults()` call.  Can  you move all of this handling over there?